build fails on python 3.x -- old print statements

Building w/ Python 3.3 fails because of old `print` statement.  

I am not sure whether this is an issue with autoconf_cig or with nemesis.

Here is the output:

```
( itchy )-( dn51drjp )-( nemesis ) autoreconf -if
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
configure.ac:23: the top level
configure.ac:18: installing './aux-config/compile'
configure.ac:11: installing './aux-config/install-sh'
configure.ac:11: installing './aux-config/missing'
Makefile.am: installing './aux-config/depcomp'
```

Then, configuring:

```
( itchy )-( dn51drjp )-( nemesis ) ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./aux-config/install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for mpicc... mpicc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether mpicc accepts -g... yes
checking for mpicc option to accept ISO C89... none needed
checking whether mpicc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of mpicc... gcc3
checking for a Python interpreter with version >= 2.3... python
checking for python... /Users/itchy/src/anaconda/bin/python
checking for python version... 3.3
checking for python platform... darwin
checking for python script directory... ${prefix}/lib/python3.3/site-packages
checking for python extension module directory... ${exec_prefix}/lib/python3.3/site-packages
checking python config... configure: error: "failed

Run '/Users/itchy/src/anaconda/bin/python python-config.py' to see what went wrong.
```

```
( itchy )-( dn51drjp )-( nemesis ) python python-config.py
  File "python-config.py", line 11
    print 'PYTHON_CPPFLAGS="%s"' % ' '.join(cppflags)
                               ^
SyntaxError: invalid syntax
```

If I switch to Python 2.7.7 I get past this point (and fail for a different reason, but that's an issue for another day).

Richard,

PyLith and none of its CIG depedencies (pyre/pythia, nemesis, 
spatialdata) have been updated for Python 3.

Since we have many Python dependencies, we have been holding off until 
we are confident in migrating. For now, use Python 2.7.

Brad

On 6/25/14, 3:29 PM, Richard Styron wrote:

> Building w/ Python 3.3 fails because of old |print| statement.
> 
> I am not sure whether this is an issue with autoconf_cig or with nemesis.
> 
> Here is the output:
> 
> |( itchy )-( dn51drjp )-( nemesis ) autoreconf -if
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:23: warning: AC_CACHE_VAL(PYTHON_INCDIR, ...): suspicious cache-id, must contain _cv_ to be cached
> ../../lib/autoconf/general.m4:2046: AC_CACHE_VAL is expanded from...
> ../../lib/autoconf/general.m4:2059: AC_CACHE_CHECK is expanded from...
> m4/cit_python.m4:20: CIT_PYTHON_INCDIR is expanded from...
> m4/cit_python.m4:35: CIT_CHECK_PYTHON_HEADER is expanded from...
> configure.ac:23: the top level
> configure.ac:18: installing './aux-config/compile'
> configure.ac:11: installing './aux-config/install-sh'
> configure.ac:11: installing './aux-config/missing'
> Makefile.am: installing './aux-config/depcomp'
> |
> 
> Then, configuring:
> 
> |( itchy )-( dn51drjp )-( nemesis ) ./configure
> checking for a BSD-compatible install... /usr/bin/install -c
> checking whether build environment is sane... yes
> checking for a thread-safe mkdir -p... ./aux-config/install-sh -c -d
> checking for gawk... no
> checking for mawk... no
> checking for nawk... no
> checking for awk... awk
> checking whether make sets $(MAKE)... yes
> checking whether make supports nested variables... yes
> checking for mpicc... mpicc
> checking whether the C compiler works... yes
> checking for C compiler default output file name... a.out
> checking for suffix of executables...
> checking whether we are cross compiling... no
> checking for suffix of object files... o
> checking whether we are using the GNU C compiler... yes
> checking whether mpicc accepts -g... yes
> checking for mpicc option to accept ISO C89... none needed
> checking whether mpicc understands -c and -o together... yes
> checking for style of include used by make... GNU
> checking dependency style of mpicc... gcc3
> checking for a Python interpreter with version >= 2.3... python
> checking for python... /Users/itchy/src/anaconda/bin/python
> checking for python version... 3.3
> checking for python platform... darwin
> checking for python script directory... ${prefix}/lib/python3.3/site-packages
> checking for python extension module directory... ${exec_prefix}/lib/python3.3/site-packages
> checking python config... configure: error: "failed
> 
> Run '/Users/itchy/src/anaconda/bin/python python-config.py' to see what went wrong.
> |
> 
> |( itchy )-( dn51drjp )-( nemesis ) python python-config.py
>    File "python-config.py", line 11
>      print 'PYTHON_CPPFLAGS="%s"' % ' '.join(cppflags)
>                                 ^
> SyntaxError: invalid syntax
> |
> 
> If I switch to Python 2.7.7 I get past this point (and fail for a
> different reason, but that's an issue for another day).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/geodynamics/nemesis/issues/1.

--- a/configure.ac
+++ b/configure.ac
@@ -18,7 +18,7 @@ AC_ARG_VAR(PYTHONPATH, [Python module search path])
 AC_PROG_CC([mpicc hcc mpcc mpcc_r mpxlc cmpicc gcc cc cl icc ecc pgcc xlc xlc_r])
 
 # Check for Python.
-AM_PATH_PYTHON([2.3])
+AM_PATH_PYTHON([3.6])
 CIT_PYTHON_CONFIG
 CIT_CHECK_PYTHON_HEADER
 CIT_CHECK_PYTHON_SHARED
@@ -28,7 +28,7 @@ CIT_CHECK_PYTHON_SHARED
 CIT_PYTHON_MODULE([pyre],[0.8.1])
 
 # Add pyre path to PYTHON_EGG_CPPFLAGS
-eval p=`$PYTHON -c "import pyre; print pyre.__""file__.replace('pyre/__init__.pyc','')"`
+eval p=`$PYTHON -c "import pyre; print(pyre.__""file__.replace('pyre/__init__.py',''))"`
 PYTHON_EGG_CPPFLAGS="-I$p"
 AC_SUBST(PYTHON_EGG_CPPFLAGS)
 

--- a/configure.ac
+++ b/configure.ac
@@ -3,7 +3,7 @@ dnl Process this file with autoconf to produce a configure script.
 dnl
 
 AC_PREREQ(2.59)
-AC_INIT([nemesis], [1.1.0], [leif@geodynamics.org])
+AC_INIT([nemesis], [2.0.0], [leif@geodynamics.org])
 AC_CONFIG_AUX_DIR([./aux-config])
 AC_CONFIG_SRCDIR([nemesis.c])
 AC_CONFIG_HEADER([config.h])
@@ -25,7 +25,7 @@ CIT_CHECK_PYTHON_SHARED
 
 # Check for Python modules and packages.
 # Assume that if we have pyre, we have pythia w/mpi.
-CIT_PYTHON_MODULE([pyre],[0.8.1])
+CIT_PYTHON_MODULE([pyre],[0.9.0])
 
 # Add pyre path to PYTHON_EGG_CPPFLAGS
 eval p=`$PYTHON -c "import pyre; print(pyre.__""file__.replace('pyre/__init__.py',''))"`

--- a/nemesis.c
+++ b/nemesis.c
@@ -63,11 +63,42 @@
 #include "mpi/_mpi.c"
 
 struct _inittab inittab[] = {
-    { "_mpi", init_mpi },
+    { "_mpi", PyInit__mpi },
     { 0, 0 }
 };
 
-int main(int argc, char **argv)
+
+void
+freeWchar(wchar_t** strings,
+	  const int nstrings) {
+  for (int i = 0; i < nstrings; ++i) {
+    PyMem_RawFree(strings[i]);
+  }
+  PyMem_Del(strings);
+} 
+  
+
+wchar_t**
+wcharFromChar(char* strings[],
+	      const int nstrings) {
+  wchar_t** wstrings = PyMem_New(wchar_t*, nstrings);
+  if (!wstrings) {
+    return NULL;
+  }
+
+  for (int i = 0; i < nstrings; ++i) {
+    wstrings[i] = Py_DecodeLocale(strings[i], NULL);
+    if (!wstrings[i]) {
+      freeWchar(wstrings, i);
+      return NULL;
+    }
+  }
+
+  return wstrings;
+}
+
+
+int main(int argc, char* argv[])
 {
     int status;
     
@@ -84,25 +115,35 @@ int main(int argc, char **argv)
         fprintf(stderr, "%s: PyImport_ExtendInittab failed! Exiting...\n", argv[0]);
         return 1;
     }
+
+    wchar_t** _argv = wcharFromChar(argv, argc);
+    if (!_argv) {
+        fprintf(stderr, "%s: Decoding argv strings failed! Exiting...\n", argv[0]);
+        return 1;
+    }
     
     if (argc < 3 || strcmp(argv[1], "--pyre-start") != 0) {
-        return Py_Main(argc, argv);
+      status = Py_Main(argc, _argv);
+      freeWchar(_argv, argc);
+      return status;
     }
     
     /* make sure 'sys.executable' is set to the path of this program  */
-    Py_SetProgramName(argv[0]);
+    Py_SetProgramName(_argv[0]);
     
     /* initialize Python */
     Py_Initialize();
     
     /* initialize sys.argv */
-    PySys_SetArgv(argc - 1, argv + 1);
+    PySys_SetArgv(argc - 1, _argv + 1);
+
+    freeWchar(_argv, argc);
     
     /* run the Python command */
     status = PyRun_SimpleString(COMMAND) != 0;
     
     /* shut down Python */
-    Py_Finalize();
+    Py_FinalizeEx();
     
 #ifdef USE_MPI
     /* shut down MPI */

